src/retrieve-validation/browser.ts

```diff
@@ -2,8 +2,31 @@
 import { W3CCSSValidatorResponse } from '.';
 
 // Utility function for retrieving response from W3C CSS Validator in a browser environment
-const retrieveInBrowser = async (url: string): Promise<W3CCSSValidatorResponse['cssvalidation']> => {
-	const res = await fetch(url);
+const retrieveInBrowser = async (url: string, timeout: number): Promise<W3CCSSValidatorResponse['cssvalidation']> => {
+	const controller = new AbortController();
+
+	if (timeout !== undefined) {
+		setTimeout(() => {
+			controller.abort();
+		}, timeout);
+	}
+
+	let res: Response | null = null;
+
+	try {
+		res = await fetch(url, { signal: timeout !== undefined ? controller.signal : undefined });
+	} catch (err) {
+		if ((err as Error).name === 'AbortError') {
+			throw new Error(`The request took longer than ${timeout}ms`);
+		}
+
+		throw new Error(err);
+	}
+
+	if (!res) {
+		throw new Error('Response expected');
+	}
+
 	const data = (await res.json()) as W3CCSSValidatorResponse;
 
 	return data.cssvalidation;
```

Comment:
Can we add some commenting in here to describe what's going on and why? There's now a lot more code in this file than before.